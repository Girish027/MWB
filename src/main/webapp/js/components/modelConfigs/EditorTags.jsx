import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { WithContext as ReactTags } from 'react-tag-input';

class EditorTags extends Component {
  constructor(props, context) {
    super(props, context);

    this.handleDelete = this.handleDelete.bind(this);
    this.handleAddition = this.handleAddition.bind(this);
    this.convertArrayToTags = this.convertArrayToTags.bind(this);

    this.props = props;

    const listItems = this.props.transformationItem ? this.props.transformationItem : [];

    this.state = {
      listItems,
    };
  }

  static getDerivedStateFromProps(nextProps) {
    return ({
      listItems: nextProps.transformationItem ? nextProps.transformationItem : [],
    });
  }

  convertArrayToTags(listItems) {
    // listItems.sort();
    const tags = [];

    if (listItems) {
      listItems.forEach((item, index) => {
        const tagItem = {
          text: item,
          id: `${index}`,
          key: `${index}`,
        };
        tags.push(tagItem);
      });
    }
    return tags;
  }

  handleDelete(index) {
    const listItems = this.state.listItems.slice();
    listItems.splice(index, 1);
    this.props.onUpdateTags(listItems);
  }

  handleAddition(tag) {
    const listItems = this.state.listItems.slice();
    listItems.push(tag.text);
    this.props.onUpdateTags(listItems);
  }

  render() {
    const tags = this.convertArrayToTags(this.state.listItems);
    const { modelViewReadOnly } = this.props;

    return (
      <div id="EditorTags">
        <ReactTags
          tags={tags}
          suggestions={[]}
          handleDelete={this.handleDelete}
          handleAddition={this.handleAddition}
          inline={false}
          readOnly={modelViewReadOnly}
        />

        {(modelViewReadOnly && tags.length === 0)
          ? <span>(No stemming exceptions defined)</span>
          : (null)
        }
      </div>
    );
  }
}

EditorTags.defaultProps = {
  onUpdateTags: () => {},
};

EditorTags.propTypes = {
  transformationItem: PropTypes.array,
  onUpdateTags: PropTypes.func,
  modelViewReadOnly: PropTypes.bool,
};

export default EditorTags;
