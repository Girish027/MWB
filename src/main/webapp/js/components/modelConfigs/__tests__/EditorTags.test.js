import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import EditorTags from 'components/modelConfigs/EditorTags';

describe('<EditorTags />', () => {
  describe('Creating an instance', () => {
    let wrapper;

    let props = {
      transformationItem: ['first', 'second', 'third'],
    };

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('should exist', () => {
      wrapper = shallow(<EditorTags {...props} />);
      expect(wrapper.exists()).toBe(true);
    });

    test('matches snapshot', () => {
      wrapper = shallow(<EditorTags {...props} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('handle addition should update parent', () => {
      props.onUpdateTags = jest.fn();
      wrapper = shallow(<EditorTags {...props} />);
      const itemToAdd = {
        id: '0',
        text: 'flower',
      };
      wrapper.instance().handleAddition(itemToAdd);
      expect(props.onUpdateTags).toHaveBeenCalledWith([...props.transformationItem, itemToAdd.text]);
    });

    test('handle deletion should update parent', () => {
      props.onUpdateTags = jest.fn();
      wrapper = shallow(<EditorTags {...props} />);
      const itemToDelete = {
        id: 2,
        text: 'third',
      };
      wrapper.instance().handleDelete(itemToDelete.id);
      expect(props.onUpdateTags).toHaveBeenCalledWith(['first', 'second']);
    });
  });
});
