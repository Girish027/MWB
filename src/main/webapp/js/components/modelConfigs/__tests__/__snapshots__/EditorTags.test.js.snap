// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditorTags /> Creating an instance matches snapshot 1`] = `
<div
  id="EditorTags"
>
  <DragDropContext(ReactTags)
    handleAddition={[Function]}
    handleDelete={[Function]}
    inline={false}
    suggestions={Array []}
    tags={
      Array [
        Object {
          "id": "0",
          "key": "0",
          "text": "first",
        },
        Object {
          "id": "1",
          "key": "1",
          "text": "second",
        },
        Object {
          "id": "2",
          "key": "2",
          "text": "third",
        },
      ]
    }
  />
</div>
`;
